addTimeout function on defer.Deferred

cyli opened the pull request:
Original ticket:  https://twistedmatrix.com/trac/ticket/5786

cyli committed:
Add timeoutDeferred and tests as it was from github.com/rackerlabs/otter/util/deferredutils,
modulo coding standard changes.

Signed-off-by: cyli <cyli@twistedmatrix.com>

cyli committed:
timeoutDeferred takes a translation function to translate cancellations into a timed out result.
By default, it translates CancelledErrors to TimedOutErrors.

Signed-off-by: cyli <cyli@twistedmatrix.com>

cyli committed:
Add docs and topfile for timeoutDeferred, and add a since tag for the new API

Signed-off-by: cyli <cyli@twistedmatrix.com>

adiroiban commented on twisted/test/test_task.py:
```diff
@@ -1199,3 +1199,197 @@ def main(reactor):
+
+        # The timeout never happens - no errback occurs
+        clock.advance(15)
+        self.assertIsNone(self.successResultOf(d))
```
this is confusing... I understand how self.successResultOf works and why we got None here, but this looks unrelated to this test.... 

so maybe this should just check that the result is not a Failure

if later successResultOf is somehow updated to preserver the result, this test will fail but no good reason :)

adiroiban commented on twisted/test/test_task.py:
```diff
@@ -1199,3 +1199,197 @@ def main(reactor):
+        d = defer.Deferred()
+        task.timeoutDeferred(d, 10, clock)
+        d.errback(_DummyException("fail"))
+        self.failureResultOf(d, _DummyException)
```
I would suggest to have 

error = _DummyException("fail")

failure = self.failureResultOf(d, _DummyException)

self.assertIs(error, failure.value)

```

this should check that the exact exception is forwarded
```

cyli replied:
Thanks!  I combined this suggestion with a modification of the test to address https://github.com/twisted/twisted/pull/145#discussion_r69741288

glyph commented:
This has conflicts that need resolving now :(

cyli committed:
Address code review comments:

1.  Renamed the callable timeoutDeferred takes to onTimeoutCancel
2.  Cleaned up the documentation so that rather than "translating CancelledErrors" we talk
    about onTimeoutCancel as just a function that gets called when the deferred is timed
    out, which can be used for logging or returning different values.
3.  Use twisted.internet.error.TimeoutError instead of creating a new error.
4.  Fix the tests to be more clear about explicitly testing that timing out doesn't happen
    if the deferred is callbacked or errbacked before the timeout.
5.  Bump the @since tag to 16.5.

Signed-off-by: cyli <cyli@twistedmatrix.com>

theisencouple commented:
FWIW, I used the following code to implement a timeout. It creates a new defer object that will callback with whatever error you want without touching the original defer. It will also allow for multiple errors and timeouts on the same defer.

class DelayError(Exception):
    pass

def delay_error(d1, exception_instance=DelayError(), wait_time=60):
    d = defer.Deferred()
    call_later = reactor.callLater(wait_time, delay_error_helper, d, exception_instance)
    d1.addCallback(delay_callback, call_later, d)
    d1.addErrback(delay_errback, call_later, d)
    return d

def delay_callback(result, call_later, d):
    try:
        call_later.cancel()
    except error.AlreadyCalled:
        pass
    if d.called:
        return
    d.callback(result)
    return result

def delay_errback(result, call_later, d):
    try:
        call_later.cancel()
    except error.AlreadyCalled:
        pass
    if d.called:
        return
    d.errback(result)
    return result

def delay_error_helper(d, exception_instance):
    if d.called:
        return
    d.errback(exception_instance)

cyli commented:
@theisencouple Thank you for the code snippet!  This seems fairly similar to this implementation, although it introduces a new deferred, since it does add a callback and errback to the original deferred.  The current implementation should also be able to handle multiple timeouts, depending on when you add the timeout.  :)  I'm updating this PR based on some feedback at a local SF meetup with some better documentation, which should hopefully help explain how to do so!

cyli committed:
Move timeoutDeferred to an addTimeout method on Deferred, and make it chainable

Signed-off-by: cyli <cyli@twistedmatrix.com>

theisencouple commented:
@cyli I thought it was important to create another defer instance so that the original defer chain was not affected by the timeout. I had previously made an implementation that used the original defer, but I was having unintended consequences (eg. the creator of the defer was not expecting the defer to be cancelled yet or a callback in the callback chain was not expecting the defer cancel error)

manishtomar commented:
> eg. the creator of the defer was not expecting the defer to be cancelled yet or a callback in the callback chain was not expecting the defer cancel error

I am sorry both of these assumptions are incorrect. `Deferred` API expects any deferred to be cancelled at any time. It is up to the creator to provide a function to _handle_ the cancellation. See http://twisted.readthedocs.io/en/twisted-16.4.1/core/howto/defer.html#cancellation

theisencouple commented:
@manishtomar I understand the underlying defer can handle a cancel and so can the defer chain. I could have used more accurate wording in my previous comment. Let me give you a scenario where I was having unintended consequences. I had a defer object that would send a message and on an error it would resend the message (or some other logic). Something like the following:
d = send_message()
d.addErrback(retry_message)

I then added code that would inform the user (eg. another process) the message was sent and added a timeout.
d.addCallback(inform_user)
reactor.callLater(timeout, d.cancel)

The problem was the user timeout was different than the underlying timeout. So when the system was experiencing load, the user timeouts would start to fire, which would cause more load (because of the retries), which caused more user timeouts to fire, and I would get into this cycle of increased load. This lead me to separate the timeouts from each other (via a new defer object) so that they would not be tied to each other and possible interact in unintended ways.

cyli commented:
@theisencouple I think I see what you're saying, but I think that you can do that with a layer built upon the `addTimeout` (which should probably affect the `Deferred` that the timeout is added to, since you'd want to actually be able to use this to time out the system ones too if you're the one implementing the system ones :)).  For example, I think this (using the existing implementation) accomplishes your goal of not canceling your system deferred:

``` python
from __future__ import print_function
from twisted.internet import task, defer


def log_call(result, label, start_time, clock):
    print("{0} called after {1} seconds with result {2!r}".format(
        label, int(clock.seconds()-start_time), result))
    return result


def main(reactor):
    start_time = reactor.seconds()

    system_d = defer.Deferred()
    user_d = defer.Deferred()

    def dontChangeSystem(result):
        user_d.callback(result)
        return result

    system_d.addBoth(dontChangeSystem)
    user_d.addTimeout(3, reactor)

    system_d.addBoth(log_call, "system_d", start_time, reactor)
    user_d.addBoth(log_call, "user_d", start_time, reactor)

    task.deferLater(reactor, 10, system_d.callback, "system one done!")
    return defer.DeferredList([system_d, user_d], consumeErrors=True)

task.react(main)

# Output:
# user_d called after 3 seconds with result <twisted.python.failure.Failure twisted.internet.defer.TimeoutError: (3, 'Deferred')>
# system_d called after 10 seconds with result 'system one done!'
```

theisencouple commented:
@cyli I think that would be a great way to separate the timeouts using the addTimeout option! As a side note, one of reason I pasted the code was in the original thread @glyph made the comment "So wait - are you saying that the timeout will be creating a new Deferred for each timeout, and ... chaining the previous Deferred in? That's interesting. That could reliably give you different exceptions, by having a custom canceler. I hadn't thought of that." I had been using some code that did something like that and thought it might be helpful to share that with you.

manishtomar commented on docs/core/howto/defer.rst:
```diff
@@ -456,8 +457,75 @@ Now if someone calls ``cancel()`` on the ``Deferred`` returned from ``HTTPClient
+:api:`twisted.internet.defer.Deferred.addTimeout <Deferred.addTimeout>` uses the :api:`twisted.internet.defer.Deferred.cancel <Deferred.cancel>` function under the hood, but can distinguish between a user's call to :api:`twisted.internet.defer.Deferred.cancel <Deferred.cancel>` and a cancellation due to a timeout.
+By default, :api:`twisted.internet.defer.Deferred.addTimeout <Deferred.addTimeout>` translates a :api:`twisted.internet.defer.CancelledError <CancelledError>` produced by the timeout into a :api:`twisted.internet.error.TimeoutError <TimeoutError>`.
+
+However, if you provided a custom :ref:`cancellation <core-howto-defer-deferreds-cancellation>` when creating the :api:`twisted.internet.defer.Deferred <Deferred>`, then cancelling it may not produce a :api:`twisted.internet.defer.CancelledError <CancelledError>`.  In this case, the default behavior of :api:`twisted.internet.task.timeoutDeferred <timeoutDeferred>` is to preserve whatever callback ...
```
Please change api link to `addTimeout` in "In this case, the default behavior of..." sentence.

cyli replied:
👍 Good spot, thanks!

cyli replied:
Fixed

cyli commented:
@theisencouple Ah thanks for the background, and your snippet!  :)  And I will add the above example to the docs.

manishtomar commented on src/twisted/internet/defer.py:
```diff
@@ -340,6 +340,67 @@ def addBoth(self, callback, *args, **kw):
+            translate a L{CancelledError} L{Failure} into a L{TimeoutError}.
+        @type onTimeoutCancel: L{callable}
+
+        @rtype: L{None}
```
Sorry another change: This function returns `self` and should document that instead of `None`. I didn't think of using this as `d.addCallback(..).addTimeout(...).addErrback(...)` until I saw the ticket.

cyli replied:
@manishtomar Thanks!  I forgot to change that when I moved it from a standalone function to one on the deferred :)

cyli replied:
Fixed

glyph commented:
@cyli @theisencouple I think that the example in question here is more confusing than helpful, in the sense that there are a _lot_ of subtleties when you start combining chaining, and layering, and cancellation, and timers.

It may be worth adding something like this to the docs at _some_ point, but as part of a more comprehensive overhaul of the Deferred documentation; the goal of documentation review is to make sure the docs are adequate for people to understand the feature–and that the feature makes enough sense to bother adding–not to block the introduction of a feature until all possible nuances and use-cases can have recipes provided for them.  This ticket is affecting something really core, so it's good that we've had a lot of discussion, but even so, we're at risk of gilding the lily by adding so much verbiage to the docs to cover it :).

If we did add such an example I really think it should have a strong motivating use-case, part of a clear narrative of _why_ you'd want to get some specific behavior out of timeouts; particularly, we'd want to have some hypothetical function other than `main` and it should have a meaningful result that the user can inspect.

cyli commented:
@glyph Ok, in that case this PR is ready for review (I think it is still marked as such on the ticket linked above)

markrwilliams commented on src/twisted/test/test_defer.py:
```diff
@@ -2446,3 +2446,301 @@ def test_cancelDeferUntilLockedWithTimeout(self):
+        d = defer.Deferred()
+        d.addTimeout(10, clock)
+
+        # timeout happens before inner deferred is returned as a errback on d
```
I'm confused by this comment.  That's because I _think_ it's the case that the timeout happens _after_ `innerDeferred` is returned from a callback\* on `d`, and consequently the test verifies that the timeout on `d` doesn't cancel `innerDeferred`.  If that's right it'd be good to change this comment.  Please don't delete it though - the comments in these tests are very helpful!

*not an errback, though that's pushing the pedantry envelope.

cyli replied:
Ah sorry, I guess I meant to say "the timeout is added before...".  You're right, the timing out actually happens afterward.  Will fix, thanks!

markrwilliams commented on src/twisted/internet/defer.py:
```diff
@@ -340,6 +340,68 @@ def addBoth(self, callback, *args, **kw):
+
+        @since: 16.5
+        """
+        timedOut = [False]
```
It's great that this isn't a new instance attribute.

markrwilliams commented on src/twisted/test/test_defer.py:
```diff
@@ -2446,3 +2446,301 @@ def test_cancelDeferUntilLockedWithTimeout(self):
+    def test_successResultBeforeTimeoutCustom(self):
+        """
+        The L{defer.Deferred} callbacks with the result if it succeeds before
+        the timeout, even if a custon C{onTimeoutCancel} function is provided.
```
`custom`, not `custon`

markrwilliams commented on src/twisted/test/test_defer.py:
```diff
@@ -2446,3 +2446,301 @@ def test_cancelDeferUntilLockedWithTimeout(self):
+        d = defer.Deferred()
+        d.addTimeout(10, clock)
+
+        # timeout happens before inner deferred is returned as a errback on d
```
I think this should read something like:
`# d is canceled before innerDeferred is returned from an errback on d`

markrwilliams commented on src/twisted/internet/defer.py:
```diff
@@ -340,6 +340,68 @@ def addBoth(self, callback, *args, **kw):
+                return toCall(value, timeout)
+            return value
+
+        self.addBoth(convertCancelled)
```
I wasn't sure why this should be `addBoth` instead of `addErrback`.  But after some deep thought, I think the core reason is that you can add `callbacks` and `errbacks` to the deferred _before_ you call `addTimeout`, and the timeout implementation has to do the Right Thing.  The Right Thing, I think, breaks down into two cases:
1. You added an errback that suppresses the `CancellationError` before `convertCancelled` fired, but you also provided `onTimeoutCancel`.  You certainly want `onTimeoutCancel` to run.  So this needs to be present in the callback chain.
2. You provided a cancellation function to the deferred that calls it back, but you also provided `onTimeoutCancel`.  You certainly want `onTimeoutCancel` to run here, too.

I went ahead and wrote some tests for my own understanding.  Maybe some of them would be useful to include in `DeferredAddTimeoutTests`.

``` python
    def test_errbackAddedBeforeTimeout(self):
        """
        An errback added before a timeout is added errbacks with a
        L{defer.CancelledError} when the timeout fires.  If the
        errback returns the L{defer.CancelledError}, it is translated
        to a L{defer.TimeoutError} by the timeout implementation.
        """
        clock = Clock()
        d = defer.Deferred()

        dErrbacked = [None]

        def errback(f):
            dErrbacked[0] = f
            return f

        d.addErrback(errback)
        d.addTimeout(10, clock)

        clock.advance(15)

        self.assertIsInstance(dErrbacked[0], failure.Failure)
        self.assertIsInstance(dErrbacked[0].value, defer.CancelledError)

        self.failureResultOf(d, defer.TimeoutError)


    def test_errbackAddedBeforeTimeoutSuppressesCancellation(self):
        """
        An errback added before a timeout is added errbacks with a
        L{defer.CancelledError} when the timeout fires.  If the
        errback suppresses the L{defer.CancelledError}, the deferred
        successfully completes.
        """
        clock = Clock()
        d = defer.Deferred()

        dErrbacked = [None]

        def errback(f):
            dErrbacked[0] = f
            f.trap(defer.CancelledError)

        d.addErrback(errback)
        d.addTimeout(10, clock)

        clock.advance(15)

        self.assertIsInstance(dErrbacked[0], failure.Failure)
        self.assertIsInstance(dErrbacked[0].value, defer.CancelledError)

        self.successResultOf(d)


    def test_errbackAddedBeforeTimeoutCustom(self):
        """
        An errback added before a timeout is added with a custom
        timeout function errbacks with a L{defer.CancelledError} when
        the timeout fires.  The timeout function runs if the errback
        returns the L{defer.CancelledError}.
        """
        clock = Clock()
        d = defer.Deferred()

        dErrbacked = [None]

        def errback(f):
            dErrbacked[0] = f
            return f

        d.addErrback(errback)
        d.addTimeout(10, clock, _overrideFunc)

        clock.advance(15)

        self.assertIsInstance(dErrbacked[0], failure.Failure)
        self.assertIsInstance(dErrbacked[0].value, defer.CancelledError)

        self.assertEqual("OVERRIDDEN", self.successResultOf(d))


    def test_errbackAddedBeforeTimeoutSuppressesCancellationCustom(self):
        """
        An errback added before a timeout is added with a custom
        timeout function errbacks with a L{defer.CancelledError} when
        the timeout fires.  The timeout function runs if the errback
        suppresses the L{defer.CancelledError}.
        """
        clock = Clock()
        d = defer.Deferred()

        dErrbacked = [None]

        def errback(f):
            dErrbacked[0] = f

        d.addErrback(errback)
        d.addTimeout(10, clock, _overrideFunc)

        clock.advance(15)

        self.assertIsInstance(dErrbacked[0], failure.Failure)
        self.assertIsInstance(dErrbacked[0].value, defer.CancelledError)

        self.assertEqual("OVERRIDDEN", self.successResultOf(d))


    def test_callbackAddedToCancelerBeforeTimeout(self):
        """
        Given a deferred with a cancellation function that resumes the
        callback chain, a callback that is added to the deferred
        before a timeout is added to runs when the timeout fires.  The
        deferred completes successfully, without a
        L{defer.TimeoutError}.
        """
        clock = Clock()
        success = "success"
        d = defer.Deferred(lambda d: d.callback(success))

        dCallbacked = [None]

        def callback(value):
            dCallbacked[0] = value
            return value

        d.addCallback(callback)
        d.addTimeout(10, clock)

        clock.advance(15)

        self.assertEqual(dCallbacked[0], success)

        self.assertIs(success, self.successResultOf(d))


    def test_callbackAddedToCancelerBeforeTimeoutCustom(self):
        """
        Given a deferred with a cancellation function that resumes the
        callback chain, a callback that is added to the deferred
        before a timeout is added to runs when the timeout fires.  The
        deferred completes successfully, without a
        L{defer.TimeoutError}.  The timeout's custom timeout function
        also runs.
        """
        clock = Clock()
        success = "success"
        d = defer.Deferred(lambda d: d.callback(success))

        dCallbacked = [None]

        def callback(value):
            dCallbacked[0] = value
            return value

        d.addCallback(callback)
        d.addTimeout(10, clock, onTimeoutCancel=_overrideFunc)

        clock.advance(15)

        self.assertEqual(dCallbacked[0], success)

        self.assertEqual("OVERRIDDEN", self.successResultOf(d))
```

cyli replied:
Wow, these are awesome, thank you!  Adding them.

markrwilliams reviewed (commented):
This is great!  The implementation looks rock solid and agrees with intuition about how it should behave.

My review consists of couple of mostly superficial suggestions.  I'm going to mark this as reviewed but feel free to re-add it.  This is a big enough change that it probably should have a few more eyes on it.

We should also have a separate issue to improve the Deferred documentation so we don't forget [@glyph's very nice suggestion](https://github.com/twisted/twisted/pull/145#issuecomment-250390306).

Thanks!  I can't wait to use this in 16.5.

glyph commented:
@markrwilliams Speaking of gilding the lily, I think we need to be unafraid to give things a passing review if they are, to the best of our ability to tell, OK :).  Nebulously calling for other reviewers tends to stall things unnecessarily, and this has already been discussed pretty intensely for literally _years_... as well as being dissected in in-person meetings, on the mailing list... there's no shortage of feedback :)

markrwilliams commented:
@glyph Fair enough :) The implementation is good and the documentation sufficient.  I would like at least some of the tests I put into [that comment](https://github.com/twisted/twisted/pull/145#discussion_r81226759) included in this PR to elucidate the interaction between timeouts and existing callbacks and errbacks.  They're merely nice to have, though, so if I don't see them here in the next 24 hours I'll merge this and open a new PR to add them.

cyli committed:
Add more tests as per review comments

Signed-off-by: cyli <cyli@twistedmatrix.com>

cyli commented:
@markrwilliams Thanks for your review!  I've added your suggested tests and hopefully rephrased the comments better.  I haven't put it back into review, as you had mentioned that you were going to merge this after 24 hours anyway, but I was hoping you could have one last look to see if the comments were more clearer.  :)

markrwilliams commented:
@cyli The latest diff looks fantastic! Now that the builders are green I'm going to merge this.

manishtomar commented:
Yay! Thank you so much for this @cyli and @markrwilliams :)

cyli commented:
@manishtomar Thank you for correcting all the doc errors!  :)
